<launch>

  <arg name="model" default="burger" doc="model type [burger, waffle, waffle_pi]"/>
<!--   <arg name="x_pos" default="2.0"/>
  <arg name="y_pos" default="-4.0"/>
  <arg name="z_pos" default="0.0"/> -->
  <arg name="x_pos" default="0.0"/>
  <arg name="y_pos" default="0.0"/>
  <arg name="z_pos" default="0.0"/>
  <!-- <arg name="yaw_init" default="3.14"/> -->
  <arg name="yaw_init" default="0.0"/>

  <arg name="base_frame" value="base_footprint"/>

  <arg name="slam_method" default="hector"/>
  <arg name="cmd_vel_topic" default="/cmd_vel" />
  <arg name="move_forward_only" default="true"/>

  <arg name="odom_topic" default="odom" />
  <arg name="odom_gazebo_topic" default="ground_truth" />
  <arg name="got_topic" default="got_pose"/>
  <arg name="odometry_filtered_topic" default="odometry/filtered_odom"/>


  <arg name="use_move_base" default="true" />
  <arg name="use_robot_localization" default="true" />
  <arg name="odom_noise_publish_tf" default="true"/> 
  <arg name="open_rviz" default="true"/> 

  <!-- Launch gazebo simulation -->
  <include file="$(find gazebo_ros)/launch/empty_world.launch">
    <!-- <arg name="world_name" value="$(find got_node)/worlds/world02_mod.world"/> -->
    <!-- <arg name="world_name" value="$(find rosbot_gazebo)/worlds/maze.world"/> -->
    <!-- <arg name="world_name" value="$(find got_node)/worlds/maze.world"/> -->
    <arg name="world_name" value="$(find got_node)/worlds/maze_mod.world"/>
    <!-- <arg name="world_name" value="$(find turtlebot3_gazebo)/worlds/turtlebot3_house.world"/> -->
    <arg name="paused" value="false"/>
    <arg name="use_sim_time" value="true"/>
    <arg name="gui" value="true"/>
    <arg name="headless" value="false"/>
    <arg name="debug" value="false"/>
  </include>


  <!-- Do no publish odom tf if robot_localization is enabled -->
  <node pkg="got_node" type="odometry_noise.py" name="odometry_with_noise" required = "true" output = "screen" >
    <param name="~old_odom_topic" value="$(arg odom_gazebo_topic)" />
    <param name="~base_frame" value="$(arg base_frame)"/>
    <param name="~odom_noise_publish_tf" value="$(arg odom_noise_publish_tf)" unless="$(arg use_robot_localization)" />
  </node>

  <!-- <node pkg="got_node" type="go_to_point.py" name="navigation"/> -->


  <!-- Launch Robot model -->
  <param name="robot_description" command="$(find xacro)/xacro --inorder $(find turtlebot3_description)/urdf/turtlebot3_$(arg model).urdf.xacro" />

  <node pkg="gazebo_ros" type="spawn_model" name="spawn_urdf"  args="-urdf -model turtlebot3_$(arg model) -x $(arg x_pos) -y $(arg y_pos) -z $(arg z_pos) -Y $(arg yaw_init) -param robot_description" />
  <!-- <node pkg="gazebo_ros" type="spawn_model" name="spawn_urdf"  args="-urdf -model $(arg model) -x $(arg x_pos) -y $(arg y_pos) -z $(arg z_pos) -Y $(arg yaw_init) -param robot_description" /> -->

  <!-- Publish location of GoT beacons -->
<!--   <node pkg="tf" type="static_transform_publisher" name="beacon1_broadcaster" args="-5 -5 5 0 0 0 1 odom beacon1 100" />
  <node pkg="tf" type="static_transform_publisher" name="beacon2_broadcaster" args="5 -5 5 0 0 0 1 odom beacon2 100" />
  <node pkg="tf" type="static_transform_publisher" name="beacon3_broadcaster" args="0 0 5 0 0 0 1 odom beacon3 100" />
  <node pkg="tf" type="static_transform_publisher" name="beacon4_broadcaster" args="-5 5 5 0 0 0 1 odom beacon4 100" />
  <node pkg="tf" type="static_transform_publisher" name="beacon5_broadcaster" args="5 5 5 0 0 0 1 odom beacon5 100" /> -->

  <!-- <include file="$(find turtlebot3_gazebo)/launch/turtlebot3_gazebo_rviz.launch" /> -->
  <!-- <include file="$(find turtlebot3_gazebo)/launch/turtlebot3_world.launch" /> -->

<!--   <node pkg="gmapping" type="slam_gmapping" name="gmapping">
      <param name="base_frame" value="base_link"/>
      <param name="odom_frame" value="odom" />
      <param name="delta" value="0.1" /<name />
  </node> -->

  <!-- Launch SLAM node  and rviz -->
  <include file="$(find turtlebot3_slam)/launch/turtlebot3_slam.launch">
    <arg name="slam_methods" value="$(arg slam_method)"/>
    <!-- <arg name="open_rviz" value="$(arg open_rviz)"/> -->
    <arg name="open_rviz" value="false"/>
  </include>


  <node pkg="got_node" type="got_position3" name="got_pos" required = "true" output = "screen" />
  <!-- Subscribed to $(arg odom_gazebo_topic) -->

  <group if="$(arg use_robot_localization)">
    <include file="$(find got_node)/launch/robot_localization.launch">
      <arg name="odom_topic" value="$(arg odom_topic)"/>
      <arg name="got_topic" value="$(arg got_topic)"/>
      <arg name="x_init" value="$(arg x_pos)"/>
      <arg name="y_init" default="$(arg y_pos)"/>
      <arg name="z_init" default="$(arg z_pos)"/>
      <arg name="yaw_init" default="$(arg yaw_init)"/>
      <arg name="odometry_filtered_topic" value="$(arg odometry_filtered_topic)"/>
    </include>
  </group>


 <!-- Navigation related -->
  <!-- Arguments -->
  <!-- <arg name="map_file" default="$(find turtlebot3_navigation)/maps/map.yaml"/> -->

  <!-- Turtlebot3  - Already initialized in turtlebot3_slam.launch-->
<!--   <include file="$(find turtlebot3_bringup)/launch/turtlebot3_remote.launch">
    <arg name="model" value="$(arg model)" />
  </include> -->


  <!-- Map server -->
  <!-- <node pkg="map_server" name="map_server" type="map_server" args="$(arg map_file)"/> -->

  <!-- AMCL -->
  <!-- <include file="$(find turtlebot3_navigation)/launch/amcl.launch"/> -->


  <!-- move_base -->
  <group if="$(arg use_move_base)"> 
    <node pkg="move_base" type="move_base" respawn="false" name="move_base" required = "true" output="screen">
      <param name="base_local_planner" value="dwa_local_planner/DWAPlannerROS" />
      <rosparam file="$(find got_node)/config/tb3burger/costmap_common_params_$(arg model).yaml" command="load" ns="global_costmap" />
      <rosparam file="$(find got_node)/config/tb3burger/costmap_common_params_$(arg model).yaml" command="load" ns="local_costmap" />
      <rosparam file="$(find got_node)/config/tb3burger/local_costmap_params.yaml" command="load" />
      <rosparam file="$(find got_node)/config/tb3burger/global_costmap_params.yaml" command="load" />
      <rosparam file="$(find got_node)/config/tb3burger/move_base_params.yaml" command="load" />
      <rosparam file="$(find got_node)/config/tb3burger/dwa_local_planner_params_$(arg model).yaml" command="load" />
      <!-- <remap from="cmd_vel" to="$(arg cmd_vel_topic)"/> -->
      <remap from="odom" to="$(arg odometry_filtered_topic)"/>
      <param name="DWAPlannerROS/min_vel_x" value="0.0" if="$(arg move_forward_only)" />
    </node>
  </group>

  <!-- ROSbot move_base configuration -->
<!--   <node pkg="move_base" type="move_base" name="move_base" output="screen">
      <param name="controller_frequency" value="10.0"/>
      <rosparam file="$(find got_node)/config/costmap_common_params.yaml" command="load" ns="global_costmap" />
      <rosparam file="$(find got_node)/config/costmap_common_params.yaml" command="load" ns="local_costmap" />
      <rosparam file="$(find got_node)/config/local_costmap_params.yaml" command="load" />
      <rosparam file="$(find got_node)/config/global_costmap_params.yaml" command="load" />
      <rosparam file="$(find got_node)/config/trajectory_planner.yaml" command="load" />
  </node> -->

  <!-- rviz -->
  <group if="$(arg open_rviz)"> 
    <node pkg="rviz" type="rviz" name="rviz" required="true"
          args="-d $(find turtlebot3_navigation)/rviz/turtlebot3_navigation.rviz"/>
          <!-- args="-d $(find turtlebot3_slam)/rviz/turtlebot3_$(arg slam_methods).rviz"/> -->
  </group>

    <node name="plotjuggler_with_layout" pkg="plotjuggler" type="PlotJuggler" required = "false" output = "screen"
          args="--nosplash --layout $(find got_node)/config/plotjuggler_layout.xml"/>

</launch>